src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.Ide.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.CodingConventions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Options
+{
+    internal sealed partial class EditorConfigDocumentOptionsProvider
+    {
+        private static readonly ConditionalWeakTable<CodingConventionsChangedEventArgs, HashSet<ProjectId>> s_projectNotifications =
+            new ConditionalWeakTable<CodingConventionsChangedEventArgs, HashSet<ProjectId>>();
+
+        partial void OnCodingConventionContextCreated(DocumentId documentId, ICodingConventionContext context)
+        {
+            context.CodingConventionsChangedAsync += (sender, e) => HandleCodingConventionsChangedAsync(documentId, sender, e);
+        }
+
+        private Task HandleCodingConventionsChangedAsync(DocumentId documentId, object sender, CodingConventionsChangedEventArgs e)
+        {
+            var projectId = documentId.ProjectId;
+            var projectsAlreadyNotified = s_projectNotifications.GetOrCreateValue(e);
+            lock (projectsAlreadyNotified)
+            {
+                if (!projectsAlreadyNotified.Add(projectId))
+                {
+                    return Task.CompletedTask;
+                }
+            }
+
+            var foregroundNotificationService = ((IMefHostExportProvider)_workspace.Services.HostServices).GetExports<IForegroundNotificationService>().Single().Value;
```

Comment:
can we make this assumption? so do we own all host who uses EditorConfigdocumentOptionProvider?

Reply:
I'm not a huge fan of this line; open to alternatives provided it doesn't add a constructor dependency on `IForegroundNotificationService` since the type isn't available in OOP.

Reply:
I just checked only one who uses EditorCOnfigDocumentOptionsProvider is VSWorkspace and remoteWorkspace so we can rely on the implementatino detail that HostService will be IMefHostExportProvider.